src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -612,4 +613,73 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         //       profile? How is this controlled at the CLI layer?
         &self.profiles.dev
     }
+
+    // Acquire extra flags to pass to the compiler from the
+    // RUSTFLAGS environment variable and similar config values
+    pub fn rustflags_args(&self, unit: &Unit) -> CargoResult<Vec<String>> {
+        // We *want* to apply RUSTFLAGS only to builds for the
+        // requested target architecture, and not to things like build
+        // scripts and plugins, which may be for an entirely different
+        // architecture. Cargo's present architecture makes it quite
+        // hard to only apply flags to things that are not build
+        // scripts and plugins though, so we do something more hacky
+        // instead to avoid applying the same RUSTFLAGS to multiple targets
+        // arches:
+        //
+        // 1) If --target is not specified we just apply RUSTFLAGS to
+        // all builds; they are all going to have the same target.
+        //
+        // 2) If --target *is* specified then we only apply RUSTFLAGS
+        // to compilation units with the Target kind, which indicates
+        // it was chosen by the --target flag.
+        //
+        // This means that, e.g. even if the specified --target is the
+        // same as the host, build scripts in plugins won't get
+        // RUSTFLAGS.
+        //
+        // Similarly, if --target is not specified we'll look in the
+        // config for the `target.rustflags` key; if it is specified
+        // we'll look in `target.$triple.rustflasg`.
+        let compiling_with_target = self.build_config.requested_target.is_some();
+        let is_target_kind = unit.kind == Kind::Target;
+
+        let use_target_rustflags = match (compiling_with_target,
+                                            is_target_kind) {
+            (false, _) => false,
+            (true, true) => true,
+            (true, false) => {
+                // This is probably a build script or plugin and we're
+                // compiling with --target. In this scenario there are
+                // no rustflags we can apply.
+                return Ok(Vec::new());
+            }
+        };
+
+        // First try RUSTFLAGS from the environment
+        if let Some(a) = env::var("RUSTFLAGS").ok() {
+            let args = a.split(" ").map(str::trim).map(str::to_string);
```

Comment:
This may want a `filter_map` for "not empty" for environment variables like "    -l foo     --sysroot bar"